typing/typecore.ml

```diff
@@ -3997,45 +3993,46 @@ and type_expect_
           sexpl types_and_modes
       in
       re {
-        exp_desc = Texp_tuple expl;
+        exp_desc = Texp_tuple (expl, alloc_mode);
         exp_loc = loc; exp_extra = [];
         (* Keep sharing *)
         exp_type = newty (Ttuple (List.map (fun e -> e.exp_type) expl));
-        exp_mode = expected_mode.mode;
         exp_attributes = sexp.pexp_attributes;
         exp_env = env }
   | Pexp_construct(lid, sarg) ->
       type_construct env expected_mode loc lid
         sarg ty_expected_explained sexp.pexp_attributes
   | Pexp_variant(l, sarg) ->
       (* Keep sharing *)
-      if sarg <> None then
-        register_allocation expected_mode;
       let ty_expected0 = instance ty_expected in
       let argument_mode = mode_subcomponent expected_mode in
       begin try match
         sarg, get_desc (expand_head env ty_expected),
         get_desc (expand_head env ty_expected0)
       with
       | Some sarg, Tvariant row, Tvariant row0 ->
+          let alloc_mode = register_allocation expected_mode in
           begin match
             row_field_repr (get_row_field l row),
             row_field_repr (get_row_field l row0)
           with
             Rpresent (Some ty), Rpresent (Some ty0) ->
               let arg = type_argument env argument_mode sarg ty ty0 in
-              re { exp_desc = Texp_variant(l, Some arg);
+              re { exp_desc = Texp_variant(l, Some (arg, alloc_mode));
                    exp_loc = loc; exp_extra = [];
                    exp_type = ty_expected0;
-                   exp_mode = expected_mode.mode;
                    exp_attributes = sexp.pexp_attributes;
                    exp_env = env }
           | _ -> raise Exit
           end
       | _ -> raise Exit
       with Exit ->
-        let arg = Option.map (type_exp env argument_mode) sarg in
-        let arg_type = Option.map (fun arg -> arg.exp_type) arg in
+        let arg = match sarg with
+        | None -> None
+        | Some sarg -> Some (type_exp env argument_mode sarg,
+                        register_allocation expected_mode)
```

Comment:
This looks like it might call `register_allocation` twice, whereas the previous design only ever called it once. Are you sure this change is safe?

Reply:
nice catch! Fixed.